libsawtooth/src/journal/genesis.rs

```diff
@@ -0,0 +1,638 @@
+    ///
+    /// GenesisError is returned if a genesis block is unable to be produced, or the resulting
+    /// block-chain-id saved.
+    pub fn start(&mut self) -> Result<(), GenesisError> {
+        let mut batch_file = File::open(self.genesis_file_path.clone()).map_err(|_| {
+            GenesisError::InvalidGenesisState(format!(
+                "Unable to open genesis batch file: {}",
+                self.genesis_file_path.display()
+            ))
+        })?;
+
+        let mut contents = vec![];
+        batch_file.read_to_end(&mut contents).map_err(|_| {
+            GenesisError::InvalidGenesisState(format!(
+                "Unable to read entire genesis batch file: {}",
+                self.genesis_file_path.display()
+            ))
+        })?;
+        let genesis_data = GenesisData::from_bytes(&contents).map_err(|_| {
+            GenesisError::InvalidGenesisData(
+                "Unable to create GenesisData from bytes in genesis batch file".to_string(),
+            )
+        })?;
+
+        let mut scheduler = self
+            .scheduler_factory
+            .create_scheduler(self.initial_state_root.to_string())
+            .map_err(|err| {
+                GenesisError::BatchValidationError(format!("Unable to create scheduler: {:?}", err))
+            })?;
+
+        let (result_tx, result_rx): (Sender<SchedulerEvent>, Receiver<SchedulerEvent>) = channel();
+        let error_tx = result_tx.clone();
+        // Add callback to convert batch result option to scheduler event
+        scheduler
+            .set_result_callback(Box::new(move |batch_result| {
+                let scheduler_event = match batch_result {
+                    Some(result) => SchedulerEvent::Result(result),
+                    None => SchedulerEvent::Complete,
+                };
+                if result_tx.send(scheduler_event).is_err() {
+                    error!("Unable to send batch result; receiver must have dropped");
+                }
+            }))
+            .map_err(|err| {
+                GenesisError::BatchValidationError(format!(
+                    "Unable to set result callback on scheduler: {}",
+                    err
+                ))
+            })?;
+
+        // add callback to convert error into scheduler event
+        scheduler
+            .set_error_callback(Box::new(move |err| {
+                if error_tx.send(SchedulerEvent::Error(err)).is_err() {
+                    error!("Unable to send scheduler error; receiver must have dropped");
+                }
+            }))
+            .map_err(|err| {
+                GenesisError::BatchValidationError(format!(
+                    "Unable to set error callback on scheduler: {}",
+                    err
+                ))
+            })?;
+
+        debug!("Adding {} genesis batches", genesis_data.batches().len());
+
+        for batch in genesis_data.batches() {
+            scheduler.add_batch(batch.clone()).map_err(|err| {
```

Comment:
It seems unnecessary to clone the batches, they should probably just be taken from the genesis data.